File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/camera/BoundOptions.java
package org.maplibre.nativeffi.camera;

import java.util.Objects;
import org.maplibre.nativeffi.geo.LatLngBounds;

/** Mutable descriptor for map camera constraint options. */
public final class BoundOptions {
  private LatLngBounds bounds;
  private Double minZoom;
  private Double maxZoom;
  private Double minPitch;
  private Double maxPitch;

  public boolean hasBounds() {
    return bounds != null;
  }

  public LatLngBounds bounds() {
    return bounds;
  }

  public BoundOptions bounds(LatLngBounds bounds) {
    this.bounds = Objects.requireNonNull(bounds, "bounds");
    return this;
  }

  public BoundOptions clearBounds() {
    bounds = null;
    return this;
  }

  public boolean hasMinZoom() {
    return minZoom != null;
  }

  public Double minZoom() {
    return minZoom;
  }

  public BoundOptions minZoom(double minZoom) {
    this.minZoom = minZoom;
    return this;
  }

  public BoundOptions clearMinZoom() {
    minZoom = null;
    return this;
  }

  public boolean hasMaxZoom() {
    return maxZoom != null;
  }

  public Double maxZoom() {
    return maxZoom;
  }

  public BoundOptions maxZoom(double maxZoom) {
    this.maxZoom = maxZoom;
    return this;
  }

  public BoundOptions clearMaxZoom() {
    maxZoom = null;
    return this;
  }

  public boolean hasMinPitch() {
    return minPitch != null;
  }

  public Double minPitch() {
    return minPitch;
  }

  public BoundOptions minPitch(double minPitch) {
    this.minPitch = minPitch;
    return this;
  }

  public BoundOptions clearMinPitch() {
    minPitch = null;
    return this;
  }

  public boolean hasMaxPitch() {
    return maxPitch != null;
  }

  public Double maxPitch() {
    return maxPitch;
  }

  public BoundOptions maxPitch(double maxPitch) {
    this.maxPitch = maxPitch;
    return this;
  }

  public BoundOptions clearMaxPitch() {
    maxPitch = null;
    return this;
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/camera/BoundOptions.kt
package org.maplibre.nativeffi.camera

import org.maplibre.nativeffi.geo.LatLngBounds

/** Mutable map bounds descriptor. */
public class BoundOptions {
  public var bounds: LatLngBounds? = null
    private set

  public var minZoom: Double? = null
    private set

  public var maxZoom: Double? = null
    private set

  public var minPitch: Double? = null
    private set

  public var maxPitch: Double? = null
    private set

  public fun hasBounds(): Boolean = bounds != null

  public fun bounds(bounds: LatLngBounds): BoundOptions = apply { this.bounds = bounds }

  public fun clearBounds(): BoundOptions = apply { bounds = null }

  public fun hasMinZoom(): Boolean = minZoom != null

  public fun minZoom(minZoom: Double): BoundOptions = apply { this.minZoom = minZoom }

  public fun clearMinZoom(): BoundOptions = apply { minZoom = null }

  public fun hasMaxZoom(): Boolean = maxZoom != null

  public fun maxZoom(maxZoom: Double): BoundOptions = apply { this.maxZoom = maxZoom }

  public fun clearMaxZoom(): BoundOptions = apply { maxZoom = null }

  public fun hasMinPitch(): Boolean = minPitch != null

  public fun minPitch(minPitch: Double): BoundOptions = apply { this.minPitch = minPitch }

  public fun clearMinPitch(): BoundOptions = apply { minPitch = null }

  public fun hasMaxPitch(): Boolean = maxPitch != null

  public fun maxPitch(maxPitch: Double): BoundOptions = apply { this.maxPitch = maxPitch }

  public fun clearMaxPitch(): BoundOptions = apply { maxPitch = null }
}
